src/pybind/mgr/diskprediction/module.py

```diff
@@ -0,0 +1,346 @@
+            'desc': 'Prints hello world to mgr.x.log',
+            'perm': 'r'
+        },
+        {
+            'cmd': 'diskprediction status',
+            'desc': 'Check diskprediction status',
+            'perm': 'r'
+        }
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(Module, self).__init__(*args, **kwargs)
+        self.status = DP_MGR_STAT_DISABLED
+        self.event = Event()
+        self.run = True
+        self._tasks = []
+        self._activate_cloud = False
+        self._prediction_result = {}
+
+    @property
+    def config_keys(self):
+        return dict((o['name'], o.get('default', None)) for o in self.OPTIONS)
+
+    def set_config_option(self, option, value):
+        if option not in self.config_keys.keys():
+            raise RuntimeError('{0} is a unknown configuration '
+                               'option'.format(option))
+
+        if option in ['diskprediction_port',
+                      'diskprediction_upload_metrics_interval',
+                      'diskprediction_upload_smart_interval',
+                      'diskprediction_retrieve_prediction_interval']:
+            try:
+                value = int(value)
+            except (ValueError, TypeError):
+                raise RuntimeError('invalid {0} configured. Please specify '
+                                   'a valid integer'.format(option))
+
+        self.log.debug('Setting in-memory config option %s to: %s', option,
+                       value)
+        self.config[option] = value
+        return True
+
+    def get_configuration(self, key):
+        return self.get_config(key, self.config_keys[key])
+
+    def _config_show(self, inbuf, cmd):
+        self.show_module_config()
+        return 0, json.dumps(self.config, indent=4), ''
+
+    def _config_mode(self, inbuf, cmd):
+        str_mode = cmd.get('mode', 'cloud')
+        if str_mode.lower() not in ['cloud', 'onpremise', 'local']:
+            return 0, 'invalid configuration, enable=[cloud|onpremise|local]', ''
+        try:
+            self.set_config('diskprediction_config_mode', str_mode)
+            return (0,
+                    'success to config disk prediction mode: %s'
+                    % str_mode.lower(), 0)
+        except Exception as e:
+            return 0, str(e), ''
+
+    def _set_smart_relied_device_health(self, inbuf, cmd):
+        str_relied = cmd.get('relied', 'false')
+        if str_relied.lower() not in ['false', 'true']:
+            return 0, 'invalid configuration, enable=[true|false]', ''
+        try:
+            self.set_config('diskprediction_smart_relied_device_health', str_relied)
+            return (0,
+                    'success to config device data %s on the device plugin'
+                    % 'not relied' if str_relied.lower() == 'false' else 'relied', 0)
+        except Exception as e:
+            return 0, 'command failed, %s' % str(e), ''
+
+    def _self_test(self, inbuf, cmd):
```

Comment:
The self test function needs to be implemented.  It should do something like gathering up the same data that we would usually send to a server, to make sure all that code works.  The idea of the self test functions is to verify that all the Ceph-facing functionality in the module works, even if we don't have an external server to send the result to.

Note that since a recent change in master, you no longer need to implement self test as a command, you can just define a function called "self_test" and it will be callable via the selftest module (see other modules for examples of self test methods).

You also need to add a function in qa/tasks/mgr/test_module_selftest.py to make sure your selftest method is called (again, you can just look at what other modules do and imitate that)